HIP/hip-1056.md

```diff
@@ -0,0 +1,1933 @@
+
+  /**
+  * BlockProof's can be in two forms either the root of the block merkle tree is directly signed in which case there is
+  * zero repeated proof_merkle_path fields. Or the BlockProof is proven based on the existance of this blocks root hash
+  * in a future block. This allows for cases where for some reason a block signature for this block could not be
+  * produced. If this is true then this will contain MerkleSiblingHash's starting from the root hash of the signed block
+  * down to and including the sibling of this blocks root hash.
+  */
+  repeated MerkleSiblingHash sibling_hashs = 5;
+}
+
+message MerkleSiblingHash {
+  /**
+  * True if this MerkleSiblingHash is the first hash in a pair of siblings, false if it is the second hash.
+  */
+  bool is_first = 1;
+
+  /**
+  * BlockHashType(SHA_384) hash for the sibling at this level of the merkle tree.
+  */
+  bytes sibling_hash = 2;
+}
+```
+
+The following diagram illustrates what the 4 sub trees contain and how blocks are chained.
+
+![Block Stream Merkle](../assets/hip-1056/block-stream-merkle.svg)
+
+The input and output item merkle trees are balanced base 2 size binary trees. The leaves across the bottom are the
+items in the order they occur in the block. If the total number of input or output items contained in the block is less
+than a power of 2 then the empty leaves are assumed to be `null` (zero length bytes). The nature of the merkle tree
+creation logic allows for the consensus node or block validator to calculate the input and output tree merkle hashes as
+items are received in a streaming way.
+
+In summary a consensus node can calculate the merkle hashes of the transaction inputs and outputs in real time and need only calculate the state merkle root at the end of a block.
+
+**Example Pseudo Code for computing streaming merkle tree root hash**
+
+```jsx
+elements = [ ... ] // array of the leaves to be streamed and hashed. Items may be transaction inputs, transaction outputs or state
+hashList = [] // a new empty list of hashes
+
+// loop over each element received and create a list of all the root hashes of the leaves of the tree 
+for (int i=0, i<elements.size(), i++) {
+    write elements[i] // output elemen to stream
+    e = hash(elements[i]) // SHA 384 item data
+    hashList.add(e) // add hash to end of list
+    
+    // for every odd located element remove and hash the last 2 hashes (representing the hash leaves) and shift right per loop to pick up the hash root
+    for (int n=i, (n & 1 == 1), n>>=1) {
+        y = hashList.getAndRemove(hashList.size() – 1); // remove last element of hashList
+        x = hashList.getAndRemove(hashList.size() – 1); // remove last element of hashList
+        hashList.add(hash(x, y)); // add the new hash root of the two hashes
+    }    
+}
+
+merkleRootHash = hashList.get(hashList.size() – 1)
+
+for (int i=hashList.size() – 2, i>=0, i--) {
+    merkleRootHash = hash(hashList.get(i), merkleRootHash)
+}
```

Comment:
This feels too simple; I'm particularly concerned about the inner loop, as it does not _seem_ likely to produce the same hash order as building the tree would.  I might be missing something, however, so perhaps @jasperpotts can help verify if this is correct.

Reply:
@derektriley confirmed this to be the case based on current services logic